src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,507 @@
+		return ErrInvalidRegion
+	}
+	auth, err := aws.GetAuth(conf.Access, conf.Secret,
+		"", time.Now().Add(s.statusTimeout))
+	if err != nil {
+		s.logger.Panic("dynamodb", "Could not log into dynamodb",
+			LogFields{"error": err.Error()})
+		return
+	}
+
+	if s.server == nil {
+		s.server = &dynamodb.Server{auth, s.region}
+	}
+	tables, err := s.server.ListTables()
+	if err != nil {
+		s.logger.Panic("dynamodb", "Could not query dynamodb",
+			LogFields{"error": err.Error()})
+		return
+	}
+	s.readProv = conf.ReadProv
+	s.writeProv = conf.WriteProv
+	s.maxChannels = conf.MaxChannels
+	s.tablename = conf.TableName
+
+	// check if the table exists
+	if !s.containsTable(tables, conf.TableName) {
+		if s.logger.ShouldLog(INFO) {
+			s.logger.Info("dynamodb", "creating dynamodb table...", nil)
+		}
+		if err = s.createTable(); err != nil {
+			s.logger.Panic("dynamodb", "Could not create table",
+				LogFields{"error": err.Error()})
+			return
+		}
+	}
+	s.pk = dynamodb.PrimaryKey{dynamodb.NewStringAttribute(UAID_LABEL, ""),
+		dynamodb.NewStringAttribute(CHID_LABEL, "")}
+	s.table = s.server.NewTable(conf.TableName, s.pk)
+	return nil
+}
+
+// CanStore indicates whether the specified number of channel registrations
+// are allowed per client. Implements Store.CanStore().
+func (s *DynamoDBStore) CanStore(channels int) bool {
+	return channels <= s.maxChannels
+}
+
+// Close closes the connection pool and unblocks all pending operations with
+// errors. Safe to call multiple times. Implements Store.Close().
+func (s *DynamoDBStore) Close() (err error) {
+	return
+}
+
+// KeyToIDs extracts the hex-encoded device and channel IDs from a user-
+// readable primary key. Implements Store.KeyToIDs().
+func (*DynamoDBStore) KeyToIDs(key string) (uaid, chid string, err error) {
+	items := strings.SplitN(key, ".", 2)
+	if len(items) < 2 {
+		return "", "", nil
+	}
+	return items[0], items[1], nil
+}
+
+// IDsToKey generates a user-readable primary key from a (device ID, channel
+// ID) tuple. The primary key is encoded in the push endpoint URI. Implements
+// Store.IDsToKey().
+func (*DynamoDBStore) IDsToKey(uaid, chid string) (string, error) {
+	if len(uaid) == 0 || len(chid) == 0 {
+		return "", nil
+	}
+	return fmt.Sprintf("%s.%s", uaid, chid), nil
+}
+
+// Status queries whether memcached is available for reading and writing.
+// Implements Store.Status().
+func (s *DynamoDBStore) Status() (success bool, err error) {
+	success = false
+	tables, err := s.server.ListTables()
+	success = err == nil && s.containsTable(tables, s.tablename)
+	return true, nil
+}
+
+// Exists returns a Boolean indicating whether a device has previously
+// registered with the Simple Push server. Implements Store.Exists().
+func (s *DynamoDBStore) Exists(uaid string) bool {
+	res, err := s.table.CountQuery([]dynamodb.AttributeComparison{
+		*dynamodb.NewEqualStringAttributeComparison(UAID_LABEL, uaid),
+		*dynamodb.NewEqualStringAttributeComparison(CHID_LABEL, " "),
```

Comment:
correct. in this case, i'm cheating by defining an illegal chid of " " as a generic. Later I'll do a search for all records with a chid > than " ", which will return all channels recorded for a UAID. It saves a few lookups.